##### Import necessary libraries

In [1]:
import numpy as np

In [2]:
import pandas as pd

In [3]:
import matplotlib as plt

In [4]:
import seaborn as sns

##### 1. Whether extend  or append?
            

In [5]:
a = [1, 2, 3, 4, 5, 6, 7, 8]  
b= [32 , 16, 21]
a.extend(b)
a

[1, 2, 3, 4, 5, 6, 7, 8, 32, 16, 21]

In [6]:
a = [1, 2 ,3 ,4, 5, 6, 7, 8]
b = [32, 16, 21]
a.append(b)
a

[1, 2, 3, 4, 5, 6, 7, 8, [32, 16, 21]]

##### My ANsWer is: a.extend(b)

####  Que 2

n.p identity(3)

#### Que 3: 

#### Lowest Average Fuel Cost per Unit Burned 

In [7]:
fue_df= pd.read_csv('fuel_ferc1.csv')

In [ ]:
fue_df.isnull().sum()

In [16]:
fue_df[['fuel_unit']] = fue_df[['fuel_unit']].fillna(value='mcf')

In [ ]:
fue_df.isnull().sum()

In [18]:
fue_df.columns

Index(['record_id', 'utility_id_ferc1', 'report_year', 'plant_name_ferc1',
       'fuel_type_code_pudl', 'fuel_unit', 'fuel_qty_burned',
       'fuel_mmbtu_per_unit', 'fuel_cost_per_unit_burned',
       'fuel_cost_per_unit_delivered', 'fuel_cost_per_mmbtu'],
      dtype='object')

In [ ]:
fue_df.groupby('fuel_type_code_pudl')[['fuel_cost_per_unit_burned']].mean()

##### Answer: Gas

#### Ques 4:

#### Standard Dev and 75th Percentile of Fuel MMBTU per Unit

In [30]:
fue_df.std(axis=0, skipna=True)

utility_id_ferc1                7.417835e+01
report_year                     7.025483e+00
fuel_qty_burned                 9.118004e+06
fuel_mmbtu_per_unit             1.060022e+01
fuel_cost_per_unit_burned       2.854490e+03
fuel_cost_per_unit_delivered    6.877593e+04
fuel_cost_per_mmbtu             2.091541e+03
dtype: float64

In [34]:
fue_df.quantile(0.75, axis=0)

utility_id_ferc1                    176.000
report_year                        2012.000
fuel_qty_burned                 1424034.000
fuel_mmbtu_per_unit                  17.006
fuel_cost_per_unit_burned            47.113
fuel_cost_per_unit_delivered         42.137
fuel_cost_per_mmbtu                   7.745
Name: 0.75, dtype: float64

##### Answer:

#### 10.60 and 17.01

#### Que 5:


#### Skewness and Kurtosis of Fuel Quantity Burned

In [39]:
fue_df.skew(axis = 0, skipna= True)

utility_id_ferc1                  0.605070
report_year                       0.006953
fuel_qty_burned                  15.851495
fuel_mmbtu_per_unit               4.135217
fuel_cost_per_unit_burned        19.787440
fuel_cost_per_unit_delivered    105.014227
fuel_cost_per_mmbtu             171.675535
dtype: float64

In [40]:
fue_df.kurt(axis= 0, skipna= True)

utility_id_ferc1                    1.088432
report_year                        -1.145656
fuel_qty_burned                   651.369450
fuel_mmbtu_per_unit                55.595695
fuel_cost_per_unit_burned         485.255851
fuel_cost_per_unit_delivered    11765.054226
fuel_cost_per_mmbtu             29489.132594
dtype: float64

#### Answer:

#### 15.85 and 651.37

#### Ques: 6

#### Answer

#### Fuel_Unit && 180 &&

###### Solution at [Cell 8]

#### Question 7:
    
######  Categorical and Mode Imputation

#### Question 8:

#### 2nd and 3rd Lowest Correlation with Fuel Cost Per Unit Burned

In [42]:
fue_df.corr(method ='pearson')

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

#### Answer

#### Fuel_qty_burned and Fuel_cost_per_unit_delivered

#### Question 9

#### Difference in Percent of Fuel Burned in 1998 and 19994

In [ ]:
new_coal= fue_df.loc[fue_df['fuel_type_code_pudl'] == 'coal']
new_coal

In [47]:
coal_year = new_coal.groupby('report_year').sum()

In [49]:
yr_a, yr_b = coal_year.loc[[1994, 1998], 'fuel_cost_per_unit_burned']

In [54]:
perc_chnge = ((yr_b - yr_a)/yr_a) * 100


In [63]:
round(perc_chnge)

-21

#### Question 10



In [ ]:
fue_df

In [66]:
fue_df.mean()

utility_id_ferc1                1.186018e+02
report_year                     2.005806e+03
fuel_qty_burned                 2.622119e+06
fuel_mmbtu_per_unit             8.492111e+00
fuel_cost_per_unit_burned       2.086490e+02
fuel_cost_per_unit_delivered    9.175704e+02
fuel_cost_per_mmbtu             1.930435e+01
dtype: float64

In [73]:
 year_fuel = fue_df.groupby(['report_year', 'fuel_cost_per_unit_delivered'])
year_fuel.first()

record_id  \
report_year fuel_cost_per_unit_delivered                             
1994        -58.800                        f1_fuel_1994_12_194_5_9   
            -35.380                       f1_fuel_1994_12_121_0_14   
             0.000                           f1_fuel_1994_12_2_1_1   
             0.027                         f1_fuel_1994_12_88_0_10   
             0.979                         f1_fuel_1994_12_151_1_7   
...                                                            ...   
2018         2956.400                        f1_fuel_2018_12_7_6_5   
             6952.161                      f1_fuel_2018_12_122_1_5   
             88419.000                     f1_fuel_2018_12_44_0_11   
             92484.000                     f1_fuel_2018_12_132_0_2   
             245652.000                    f1_fuel_2018_12_132_0_1   

                                          utility_id_ferc1  plant_name_ferc1  \
report_year fuel_cost_per_unit_delivered                                       
1994        -58.800                                    194        rock river   
            -35.380                                    121     french island   
             0.000                                       2  joseph m. farley   
             0.027                                      88          cane run   
             0.979                                     151        station 13   
...                                                    ...               ...   
2018         2956.400                                    7      palo verde 3   
             6952.161                                  122       huron gt #2   
             88419.000                                  44            monroe   
             92484.000                                 132            coyote   
             245652.000                                132            coyote   

                                         fuel_type_code_pudl fuel_unit  \
report_year fuel_cost_per_unit_delivered                                 
1994        -58.800                                    waste       ton   
            -35.380                                    waste       ton   
             0.000                                   nuclear       kgU   
             0.027                                       gas       mcf   
             0.979                                   nuclear    gramsU   
...                                                      ...       ...   
2018         2956.400                                nuclear       kgU   
             6952.161                                    oil       bbl   
             88419.000                                   oil       bbl   
             92484.000                                   oil       bbl   
             245652.000                                 coal       ton   

                                          fuel_qty_burned  \
report_year fuel_cost_per_unit_delivered                    
1994        -58.800                                4650.0   
            -35.380                               51728.0   
             0.000                                 2260.0   
             0.027                                12604.0   
             0.979                               602820.0   
...                                                   ...   
2018         2956.400                               467.0   
             6952.161                                 3.0   
             88419.000                            47073.0   
             92484.000                             6869.0   
             245652.000                          879945.0   

                                          fuel_mmbtu_per_unit  \
report_year fuel_cost_per_unit_delivered                        
1994        -58.800                                  0.014370   
            -35.380                                  0.005846   
             0.000                                   0.064094   
             0.027     

In [ ]:
fue_df

In [84]:
fue_df.describe()

utility_id_ferc1   report_year  fuel_qty_burned  fuel_mmbtu_per_unit  \
count      29523.000000  29523.000000     2.952300e+04         29523.000000   
mean         118.601836   2005.806050     2.622119e+06             8.492111   
std           74.178353      7.025483     9.118004e+06            10.600220   
min            1.000000   1994.000000     1.000000e+00             0.000001   
25%           55.000000   2000.000000     1.381700e+04             1.024000   
50%          122.000000   2006.000000     2.533220e+05             5.762694   
75%          176.000000   2012.000000     1.424034e+06            17.006000   
max          514.000000   2018.000000     5.558942e+08           341.260000   

       fuel_cost_per_unit_burned  fuel_cost_per_unit_delivered  \
count               29523.000000                  2.952300e+04   
mean                  208.649031                  9.175704e+02   
std                  2854.490090                  6.877593e+04   
min                  -276.080000                 -8.749370e+02   
25%                     5.207000                  3.778500e+00   
50%                    26.000000                  1.737100e+01   
75%                    47.113000                  4.213700e+01   
max                139358.000000                  7.964521e+06   

       fuel_cost_per_mmbtu  
count         29523.000000  
mean             19.304354  
std            2091.540939  
min             -41.501000  
25%               1.940000  
50%               4.127000  
75%               7.745000  
max          359278.000000

In [85]:
fue_df.mean(axis = 0)

utility_id_ferc1                1.186018e+02
report_year                     2.005806e+03
fuel_qty_burned                 2.622119e+06
fuel_mmbtu_per_unit             8.492111e+00
fuel_cost_per_unit_burned       2.086490e+02
fuel_cost_per_unit_delivered    9.175704e+02
fuel_cost_per_mmbtu             1.930435e+01
dtype: float64

In [114]:
fue_df

record_id  utility_id_ferc1  report_year  \
0        f1_fuel_1994_12_1_0_7                 1         1994   
1       f1_fuel_1994_12_1_0_10                 1         1994   
2        f1_fuel_1994_12_2_0_1                 2         1994   
3        f1_fuel_1994_12_2_0_7                 2         1994   
4       f1_fuel_1994_12_2_0_10                 2         1994   
...                        ...               ...          ...   
29518  f1_fuel_2018_12_12_0_13                12         2018   
29519   f1_fuel_2018_12_12_1_1                12         2018   
29520  f1_fuel_2018_12_12_1_10                12         2018   
29521  f1_fuel_2018_12_12_1_13                12         2018   
29522  f1_fuel_2018_12_12_1_14                12         2018   

           plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0                  rockport                coal       ton        5377489.0   
1      rockport total plant                coal       ton       10486945.0   
2                    gorgas                coal       ton        2978683.0   
3                     barry                coal       ton        3739484.0   
4                 chickasaw                 gas       mcf          40533.0   
...                     ...                 ...       ...              ...   
29518    neil simpson ct #1                 gas       mcf          18799.0   
29519  cheyenne prairie 58%                 gas       mcf         806730.0   
29520     lange ct facility                 gas       mcf         104554.0   
29521       wygen 3 bhp 52%                coal       ton         315945.0   
29522       wygen 3 bhp 52%                 gas       mcf          17853.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0                   16.590                      18.59   
1                   16.592                      18.58   
2                   24.130                      39.72   
3                   23.950                      47.21   
4                    1.000                       2.77   
...                    ...                        ...   
29518                1.059                       4.78   
29519                1.050                       3.65   
29520                1.060                       4.77   
29521               16.108                       3.06   
29522                1.059                       0.00   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                             18.53                1.121  
1                             18.53                1.120  
2                             38.12                1.650  
3                             45.99                1.970  
4                              2.77                2.570  
...                             ...                  ...  
29518                          4.78                9.030  
29519                          3.65                6.950  
29520                          4.77                8.990  
29521                         14.76                1.110  
29522                          0.00               11.680  

[29523 rows x 11 columns]

In [122]:
new_fue= fue_df.groupby('report_year')['fuel_cost_per_unit_delivered'].mean()
new_fue

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

##### Answer

#### 1997

In [89]:
fue_year = fue_df['report_year']

In [90]:
fue_fuel = fue_df['fuel_cost_per_unit_delivered']

In [91]:
fue_year_fuel = fue_df[['report_year', 'fuel_cost_per_unit_delivered']]

In [92]:
fue_year_fuel

report_year  fuel_cost_per_unit_delivered
0             1994                         18.53
1             1994                         18.53
2             1994                         38.12
3             1994                         45.99
4             1994                          2.77
...            ...                           ...
29518         2018                          4.78
29519         2018                          3.65
29520         2018                          4.77
29521         2018                         14.76
29522         2018                          0.00

[29523 rows x 2 columns]